cmake/helpers/GdalGenerateConfig.cmake

```diff
@@ -84,9 +144,12 @@ function(gdal_split_library_to_lflags _lib_flag _path_flag _input)
         if(_pos EQUAL "-1")
             continue()
         endif()
+        get_filename_component(_lib_dir "${_input}" DIRECTORY)
         string(SUBSTRING "${_lib_name}" 0 "${_pos}" _name_we)
         if(NOT "${_name_we}${_suffix}" STREQUAL _lib_name)
-            continue()
+            if(NOT _suffix STREQUAL ".so" OR NOT EXISTS "${_lib_dir}/${_name_we}${_suffix}")
```

Comment:
With the information from https://lists.osgeo.org/pipermail/proj/2022-February/010551.html, it might be more obvious why splitting can be a bad idea: We really don't know if `-lfoo` resolves to `/original/libfoo.so.1.2.3` for the linker and to `libfoo.so.<SOVERSION>` being added to lib gdal.

Reply:
> We really don't know if `-lfoo`

I don't think that splitting is a bad idea. When building, -lfoo resolves to finding libfoo.so, which is the development oriented file, and there must be a single one at a time in the directories you link to, and it is a symlink to the actual library libfoo.so.SONAME, and that will be the one kept in the linked library. So if you have a sane environment, linking against /path/libfoo.so.1.2.3, /path/libfoo.so.1 or -Lpath -lfoo will end up in the same result.

Reply:
IMO it is not so easy. 
- `-lfoo` prefers shared over static (unless building with `--static`). I acknowledge that static is less relevant but it should be noted nevertheless.
- > there must be a single one at a time in the directories you link to
  
  There can be only one `libfoo.so` in a single directory, but there can be one in *each directory*. In fact there must be one in each prefix *for development*: `-lfoo` is common in pkg-config files, and `libfoo.so` is also what `find_library` looks for.

- `-lfoo` must be complemented by `-L/bar`, but `-L/bar`'s effect is not scoped to `-lfoo`.

  While we can safely append to the explicit search path, the explicit path will be before the implicit search path AFAIU. One might suddenly get a different `libz.so` from `-lfoo`. 
  Basically, you already pointed to the biggest source of trouble in another comment: `find_package` looking directories derived from `PATH`. This is somewhat okay for consistent linking in CMake, with absolute paths. But it gets dangerous when we start to take the lib's directory separately and add it to the linker search path. One could say this is not how CMake config is meant to be used. 
  It may perhaps be summarized as: Don't use `-L` where CMake wouldn't use `-L`.
  
  (Even the current version of `FindPkgConfig.cmake` resolves the given `Libs` into absolute paths for use in the imported target's property.)

Reply:
I agree with your observations. I'm not sure if you are asking for specific feedback on what to do ?